daml-lf/data/src/main/scala/com/digitalasset/daml/lf/data/Struct.scala

```diff
@@ -6,43 +6,82 @@ package data
 
 import com.daml.lf.data.Ref.Name
 
-final case class Struct[+X] private (sortedFields: ImmArray[(Ref.Name, X)]) extends NoCopy {
+/** We use this container to describe structural record as sorted flat list in various parts of the codebase.
+    `entries` are sorted by their first component without duplicate.
+  */
+final case class Struct[+X] private (private val entries: ImmArray[(Ref.Name, X)]) extends NoCopy {
 
-  def lookup(name: Ref.Name): Option[X] =
-    sortedFields.find(_._1 == name).map(_._2)
+  /* Invariant:
 
-  def mapValue[Y](f: X => Y) = new Struct(sortedFields.map { case (k, v) => k -> f(v) })
+      for {
+        i <- sortedFields.indices
+        j <- sortedFields.indices
+        if (i < j)
+      } assert((sortedFields(i)._1 compare sortedFields(j)._1) < 0)
```

Comment:
Not sure the formalism is helpful here. “must be sorted by the first element in strictly ascending order” seems just as precise and at least for me it is easier to parse.

Reply:
dropped